actions/promote/createBatch.js

```diff
@@ -0,0 +1,170 @@
+                status: FgStatus.PROJECT_STATUS.FAILED,
+                statusMessage: payload
+            });
+        } else {
+            urlInfo.setUrlInfo(adminPageUri);
+            payload = 'Getting all files to be promoted.';
+            await fgStatus.updateStatusToStateLib({
+                status: FgStatus.PROJECT_STATUS.IN_PROGRESS,
+                statusMessage: payload
+            });
+            logger.info(payload);
+            payload = 'Creating batches.';
+            payload = await createBatch(batchManager);
+            await fgStatus.updateStatusToStateLib({
+                status: FgStatus.PROJECT_STATUS.IN_PROGRESS,
+                statusMessage: payload,
+                batchesInfo: batchManager.getBatchesInfo()
+            });
+            logger.info(payload);
+
+            // Finalize and Trigger N Track the batches
+            await batchManager.finalizeInstance({
+                adminPageUri, projectExcelPath, fgRootFolder, doPublish
+            });
+            logger.info('Instance finalized and started');
+        }
+    } catch (err) {
+        await fgStatus.updateStatusToStateLib({
+            status: FgStatus.PROJECT_STATUS.COMPLETED_WITH_ERROR,
+            statusMessage: err.message,
+        });
+        logger.error(err);
+        payload = err;
+    }
+
+    return {
+        body: payload,
+    };
+}
+
+/**
+ * Find all files in the pink tree to promote.
+ */
+async function findAllFiles() {
+    const { sp } = await getConfig();
+    const baseURI = `${sp.api.excel.update.fgBaseURI}`;
+    const rootFolder = baseURI.split('/').pop();
+    const options = await getAuthorizedRequestOption({ method: 'GET' });
+    // Temporarily restricting the iteration for promote to under /drafts folder only
+    return findAllFloodgatedFiles(baseURI, options, rootFolder, [], ['/drafts']);
+}
+
+/**
+ * Iteratively finds all files under a specified root folder.
+ */
+async function findAllFloodgatedFiles(baseURI, options, rootFolder, fgFiles, fgFolders) {
+    while (fgFolders.length !== 0) {
+        const uri = `${baseURI}${fgFolders.shift()}:/children?$top=${MAX_CHILDREN}`;
+        // eslint-disable-next-line no-await-in-loop
+        const res = await fetchWithRetry(uri, options);
+        if (res.ok) {
+            // eslint-disable-next-line no-await-in-loop
+            const json = await res.json();
+            // eslint-disable-next-line no-await-in-loop
+            const driveItems = json.value;
+            driveItems?.forEach((item) => {
+                const itemPath = `${item.parentReference.path.replace(`/drive/root:/${rootFolder}`, '')}/${item.name}`;
+                if (item.folder) {
+                    // it is a folder
+                    fgFolders.push(itemPath);
+                } else {
+                    const downloadUrl = item['@microsoft.graph.downloadUrl'];
+                    fgFiles.push({ fileDownloadUrl: downloadUrl, filePath: itemPath });
```

Comment:
As we are writing this data to a file, also considering the fact that the download URL is only valid for 60min, we can remove reading/writing `downloadUrl` data here. 

Instead, during the copy API call as part of promote action, if the file is locked, let's make a MS Graph API call to get the download URL of that specific file and use it there. This will avoid unnecessary read and storage of download URLs across 1000s of files.